src/main/java/io/harness/cf/client/api/MetricsProcessor.java

```diff
@@ -39,42 +34,95 @@
   private static final String SDK_LANGUAGE = "SDK_LANGUAGE";
   private static final String SDK_VERSION = "SDK_VERSION";
 
-  private String jarVerion = "";
-
-  private final MetricsApi metricsAPI;
-  private final Cache analyticsCache;
-  private final String environmentID;
-  private final String cluster;
+  private final MetricsCallback callback;
   private final Config config;
+  private final BlockingQueue<MetricEvent> queue;
+  private final MetricsApi api;
+
+  private String jarVersion = "";
+
+  @Setter private String environmentID;
+  @Setter private String cluster;
 
-  public AnalyticsPublisherService(
-      MetricsApi metricsAPI,
-      Config config,
-      String environmentID,
-      String cluster,
-      Cache analyticsCache) {
-
-    this.metricsAPI = metricsAPI;
-    this.analyticsCache = analyticsCache;
-    this.environmentID = environmentID;
-    this.cluster = cluster;
+  private String token;
+
+  public MetricsProcessor(Config config, MetricsCallback callback) {
     this.config = config;
+    this.callback = callback;
+    this.api = new MetricsApi();
+    this.queue = new LinkedBlockingQueue<>(config.getBufferSize());
+    this.callback.onMetricsReady();
+  }
+
+  public void setToken(String token) {
+    this.token = token;
+    this.api.setApiClient(makeApiClient());
+  }
+
+  protected ApiClient makeApiClient() {
+    final int maxTimeout = 30 * 60 * 1000;
+    ApiClient apiClient = new ApiClient();
+    apiClient.setBasePath(config.getEventUrl());
+    apiClient.setConnectTimeout(maxTimeout);
+    apiClient.setReadTimeout(maxTimeout);
+    apiClient.setWriteTimeout(maxTimeout);
+    apiClient.setDebugging(log.isDebugEnabled());
+    apiClient.addDefaultHeader("Authorization", "Bearer " + token);
+    return apiClient;
+  }
+
+  // push the incoming data to the ring buffer
+  public void pushToQueue(Target target, FeatureConfig featureConfig, Variation variation) {
+
+    MetricEvent analytics =
+        MetricEvent.builder()
```

Comment:
builders are discouraged unless some fields are optional (where you would explicitly require mandatory fields). Here seems that you need constructor

Reply:
make sense, it is removed :)